### DISCUSSION TITLE
[Question]: Keycloak login

### DISCUSSION THREAD

**Discussion by MarcelRei** [danny-avila/LibreChat#3182] [Troubleshooting]
### What is your question?

Hi,

I've setup Librechat on Debian12 and connected my Keycloak 23.0.6. I can login and I see a successful login in the debug log, but I just get redirected to the default login page. The Keycloakserver shows an active login aswell
Logging in with my local account, I get forwarded to the chat.

### More Details

# OpenID
OPENID_CLIENT_ID=librechat
OPENID_CLIENT_SECRET=******************
OPENID_ISSUER=https://keycloak.domain.com/realms/Realmname
OPENID_SESSION_SECRET=*********
OPENID_SCOPE="openid profile email"
OPENID_CALLBACK_URL=/oauth/openid/callback
#OPENID_REQUIRED_ROLE=
#OPENID_REQUIRED_ROLE_TOKEN_KIND=
#OPENID_REQUIRED_ROLE_PARAMETER_PATH=

OPENID_BUTTON_LABEL=JvM Login
OPENID_IMAGE_URL=

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by danny-avila** [comment]
There were a couple of people setting up keycloak recently:

https://github.com/danny-avila/LibreChat/discussions/3165
https://github.com/danny-avila/LibreChat/discussions/3073

Let me know if these are helpful

**Comment by prononext** [comment]
Maybe you could share your LibreChat logs?
I had the same problem and it worked as soon as I set a proper Realm Role and assigned it to the user/group I logged into LibreChat with. 
https://github.com/danny-avila/LibreChat/discussions/3165

**Comment by MarcelRei** [comment]
.env:
OPENID_CLIENT_ID=librechat
OPENID_CLIENT_SECRET=aaaaaaaaaaaaaaaaaaaaaaaaaaa
OPENID_ISSUER=https://keycloak.domain.de/realms/Realm
OPENID_SESSION_SECRET=helloworld
OPENID_SCOPE=openid profile email userID
OPENID_CALLBACK_URL=/oauth/openid/callback
OPENID_REQUIRED_ROLE=libregpt
OPENID_REQUIRED_ROLE_TOKEN_KIND=access
OPENID_REQUIRED_ROLE_PARAMETER_PATH=realm_access.roles

OPENID_BUTTON_LABEL=JvM Login
OPENID_IMAGE_URL=

2024-06-25T09:48:24.948Z info: [openidStrategy] verify login openidId: *************************
2024-06-25T09:48:24.949Z debug: [openidStrategy] very login tokenset and userinfo
{
    tokenset.access_token: "*************************5... [truncated]",
    tokenset.expires_at: 1719309204,
    tokenset.refresh_expires_in: 1800,
    tokenset.refresh_token: "**********************************i... [truncated]",
    tokenset.token_type: "Bearer",
    tokenset.id_token: "*****************************... [truncated]",
    tokenset.not-before-policy: 0,
    tokenset.session_state: "2311daf0-f30f-449d-a04a-f1c013e274ff",
    tokenset.scope: "openid email profile userID",
    userinfo.sub: "40fac4e8-a03f-4a81-9699-dfb88af52bab",
    userinfo.email_verified: true,
    userinfo.user_id: "marcel.******",
    // 37 ad-group(s)
    userinfo.ad-groups: [""],
    userinfo.name: "Marcel",
    userinfo.preferred_username: "marcel",
    userinfo.given_name: "Marcel",
    userinfo.family_name: "*****",
    userinfo.email: "marcel.****@domain.com",
    userinfo.username: "marcel.****",
}
2024-06-25T09:48:24.954Z info: [openidStrategy] user found with openidId: 40fac4e8-a03f-4a81-9699-dfb88af52bab
2024-06-25T09:48:24.971Z info: [openidStrategy] login success openidId: 40fac4e8-a03f-4a81-9699-dfb88af52bab | email: marcel.********@domain.com | username: marcel.****
2024-06-25T09:48:25.277Z debug: [validateJson] files
["\"Ai_PDF.json\"","\"BrowserOp.json\"","\"Dr_Thoths_Tarot.json\"","\"DreamInterpreter.json\"","\"VoxScript.json\"","\"askyourpdf.json\"","\"drink_maestro.json\"","\"earthImagesAndVisualizations.json\"","\"image_prompt_enhancer.json\"","\"qrCodes.json\"","\"scholarai.json\"","\"uberchord.json\"","\"web_search.json\""]
2024-06-25T09:48:25.286Z debug: [validateJson] files
["\"Ai_PDF.json\"","\"BrowserOp.json\"","\"Dr_Thoths_Tarot.json\"","\"DreamInterpreter.json\"","\"VoxScript.json\"","\"askyourpdf.json\"","\"drink_maestro.json\"","\"earthImagesAndVisualizations.json\"","\"image_prompt_enhancer.json\"","\"qrCodes.json\"","\"scholarai.json\"","\"uberchord.json\"","\"web_search.json\""]

**Reply by prononext** [reply]
maybe you try:

**`OPENID_SCOPE=openid profile email`**

instead of:

`OPENID_SCOPE=openid profile email userID`

and be sure the `libregpt` role is a realm role and not a client role and is assigned correctly. 

If the problem continues, maybe share some keycloak logs. 
Had had a long debugging way to get keycloak to work but finally its done and working.

**Comment by MarcelRei** [comment]
I've tried it without userID before, still doesn't work. libregpt ist a realm roll and assigned to my user. I will check the Keycloak logs and update you

**Reply by MarcelRei** [reply]
I have to say I've given up on this and just went with LDAP Authentication instead

**Comment by MarcelRei** [comment]
I don't really find anything. If setup a librechat.yaml file as well. Do I need to setup something inthere aswell?

# Configuration version (required)
version: 1.1.4

# Cache settings: Set to true to enable caching
cache: true

# Custom interface configuration
interface:
  # Privacy policy settings
  privacyPolicy:
    externalUrl: 'https://*****************'
    openNewTab: true

  # Terms of service
  termsOfService:
    externalUrl: 'https://**************'
    openNewTab: true
  endpointsMenu: true
  parameters: true
  sidePanel: true
  presets: false

# Example Registration Object Structure (optional)
registration:
  socialLogins: ['openid']

# Definition of custom endpoints
endpoints:
  azureOpenAI:
    titleModel: "gpt-4-turbo"
    plugins: true
    groups:
      - group: "eu" # arbitrary name
        apiKey: "${OPENAI_AZURE_EU_API_KEY}"
        instanceName: "gpt" # name of the resource group or instance
        version: "************"
        # baseURL: https://prod.example.com
        # additionalHeaders:
        #   X-Custom-Header: value
        models:
          gpt-3.5-turbo:
            deploymentName: gpt-35-turbo
          gpt-4:
            deploymentName: gpt-4
          gpt-4-1106-preview:
            deploymentName: gpt-4-1106-preview

I also tried a different realm in my Keycloak

**Comment by LtChaser** [comment]
For me the issue was that the user in our keycloak setup got their role assigned as a client role and not a realm role so I had to change the OPENID_REQUIRED_ROLE_PARAMETER_PATH to "resource_access.<client-id>.roles" and then it started working. Hope it helps!

**Comment by MarcelRei** [comment]
Nope, still the same issue. The role is setup as a realm roll aswell

**Comment by prononext** [comment]
Hi @MarcelRei, 

I was also struggeling to get it working with the same error. Tried tons of possibilities and finally it worked. 
**Be sure you have this setup exactly! (Following the Docs can be confusing, so I shared exactly the values I used)**

https://github.com/danny-avila/LibreChat/discussions/3165#discussioncomment-9862121

**In your Keycloak, Client Settings Only Home URL and Valid redirect URLs should be filled:**

Home URL: https://your-librechat-domain
Valid redirect URIs: https://your-librechat-domain/oauth/openid/callback

**Comment by MarcelRei** [comment]
I've given up on this and went for AD authentication instead
